src/ui/components/HostPermissions/index.js

```diff
@@ -0,0 +1,166 @@
+/* @flow */
+import * as React from 'react';
+import { compose } from 'redux';
+
+import log from 'core/logger';
+import translate from 'core/i18n/translate';
+import type { I18nType } from 'core/types/i18n';
+import Permission from 'ui/components/Permission';
+
+/* eslint-disable no-continue */
+type Props = {|
+  i18n: I18nType,
+  permissions: Array<string>,
+|};
+
+type HostPermissionMessageType = 'domain' | 'site';
+type PermissionMessageType = HostPermissionMessageType | 'allUrls';
+
+type GetPermissionStringParams = {|
+  messageType: PermissionMessageType,
+  param?: string | number,
+  tooMany?: boolean,
+|};
+
+type GenerateHostPermissionsParams = {|
+  permissions: Array<string>,
+  // eslint-disable-next-line react/no-unused-prop-types
+  messageType: HostPermissionMessageType,
+|};
+
+export class HostPermissionsBase extends React.Component<Props> {
+  constructor(props: Props) {
+    super(props);
+    this.i18n = props.i18n;
+  }
+
+  getPermissionString({ messageType, param, tooMany = false }: GetPermissionStringParams): string {
```

Comment:
All of the mention of "too many" in this component reads like an error to me, eg: "Too many permissions" feels like an error, but it's really "too many to display so we'll truncate".

Maybe this would be better named `truncatePermissions` or `multipleDomains`?